spec/datadog/profiling/collectors/discrete_dynamic_sampler_spec.rb

```diff
@@ -0,0 +1,226 @@
+        expect(stats[:overhead]).to be < max_overhead_target
+        expect(sampler._native_probability).to be_between(1, 3)
+      end
+    end
+  end
+
+  context 'when under a variable load' do
+    context 'containing lots of short spikes' do
+      it 'will readjust to decrease sampling rate' do
+        # Baseline
+        simulate_load(duration_seconds: 10, events_per_second: 10, sampling_seconds: 0.01)
+        p_baseline = sampler._native_probability
+
+        # We'll spike every 5 seconds. Sampler should have some short term memory so
+        # after a spiking period it should be using a lower probability
+        simulate_load(duration_seconds: 1, events_per_second: 1000, sampling_seconds: 0.01)
+        simulate_load(duration_seconds: 5, events_per_second: 10, sampling_seconds: 0.01)
+        simulate_load(duration_seconds: 1, events_per_second: 1000, sampling_seconds: 0.01)
+        simulate_load(duration_seconds: 5, events_per_second: 10, sampling_seconds: 0.01)
+        simulate_load(duration_seconds: 1, events_per_second: 1000, sampling_seconds: 0.01)
+        simulate_load(duration_seconds: 10, events_per_second: 10, sampling_seconds: 0.01)
+
+        p_after_spikes = sampler._native_probability
+
+        expect(p_after_spikes).to be < p_baseline
+      end
+    end
+
+    context 'with a big spike at the beginning' do
+      it "won't wait until the next sample to adjust" do
+        # We'll start with a very big load at the beginning. This should move the sampler towards
+        # having very low sampling probabilities (i.e. a big sampling interval). We want to validate
+        # that the sampler can readjust in-between samples so it can react to load pattern changes
+        # that may happen inbetween. If that weren't the case, after we go down to a very low event
+        # baseline, entire minutes could pass before we even decide to sample again and realize that
+        # we've been sampling nothing for a while.
+        simulate_load(duration_seconds: 5, events_per_second: 100000, sampling_seconds: 0.1)
+        p1 = sampler._native_probability
+        expect(p1).to be < 0.1 # %
+
+        # With such a low probability, our sampling skip is >1000 so if we relied on samples alone
+        # for adjustment, the events generated in each of the following loads would not be enough to
+        # trigger this readjustment. We expect that continuous adjustment slowly brings probabilities
+        # up and eventually we can sample again within some reasonable amount of settings (15 secs
+        last_p = p1
+        saw_sample = 10.times do
+          stats = simulate_load(duration_seconds: 2, events_per_second: 1, sampling_seconds: 0.1)
+          current_p = sampler._native_probability
+          expect(current_p).to be > last_p
+
+          last_p = current_p
+
+          break true if stats[:num_samples] > 0
+        end
+        expect(saw_sample).to be(true)
```

Comment:
Should we maybe assert on some expected rough target % after these multiple adjustment rounds?